src/annotate.cpp

```diff
@@ -0,0 +1,524 @@
+class wayfire_annotate_screen : public wf::plugin_interface_t
+{
+    wlr_box last_bbox;
+    bool hook_set = false;
+    anno_ws_overlay shape_overlay;
+    wf::pointf_t grab_point, last_cursor;
+    std::vector<std::vector<anno_ws_overlay>> overlays;
+    wf::option_wrapper_t<std::string> method{"annotate/method"};
+    wf::option_wrapper_t<double> line_width{"annotate/line_width"};
+    wf::option_wrapper_t<bool> shapes_from_center{"annotate/from_center"};
+    wf::option_wrapper_t<wf::color_t> stroke_color{"annotate/stroke_color"};
+    wf::option_wrapper_t<wf::buttonbinding_t> draw_binding{"annotate/draw"};
+    wf::option_wrapper_t<wf::activatorbinding_t> clear_binding{"annotate/clear_workspace"};
+
+    public:
+    void init() override
+    {
+        grab_interface->name = "annotate";
+        grab_interface->capabilities = wf::CAPABILITY_MANAGE_COMPOSITOR;
+
+        auto wsize = output->workspace->get_workspace_grid_size();
+        overlays.resize(wsize.width);
+        for (int x = 0; x < wsize.width; x++)
+        {
+            overlays[x].resize(wsize.height);
+        }
+
+        grab_interface->callbacks.pointer.relative_motion = [=] (wlr_event_pointer_motion* ev)
+        {
+            pointer_moved(ev);
+        };
+
+        grab_interface->callbacks.pointer.button = [=] (uint32_t b, uint32_t s)
+        {
+            if (s == WL_POINTER_BUTTON_STATE_RELEASED)
+            {
+                draw_end();
+            }
+        };
+
+        output->connect_signal("reserved-workarea", &workarea_changed);
+        output->connect_signal("viewport-changed", &viewport_changed);
+        output->add_button(draw_binding, &draw_begin);
+        output->add_activator(clear_binding, &clear_workspace);
+    }
+
+    anno_ws_overlay& get_current_overlay()
+    {
+        auto ws = output->workspace->get_current_workspace();
+        return overlays[ws.x][ws.y];
+    }
+
+    wf::signal_connection_t viewport_changed{[this] (wf::signal_data_t *data)
+    {
+        output->render->damage_whole();
+    }};
+
+    wf::button_callback draw_begin = [=] (uint32_t, int x, int y)
+    {
+        grab_point = last_cursor = wf::get_core().get_cursor_position();
+
+        grab();
+
+        return true;
+    };
+
+    void draw_end()
+    {
+        auto& ol = get_current_overlay();
+
+        overlay_destroy(shape_overlay);
+        ungrab();
+
+        if (std::string(method) == "line")
+        {
+            cairo_draw_line(ol, wf::get_core().get_cursor_position());
+        }
+        else if (std::string(method) == "rectangle")
+        {
+            cairo_draw_rectangle(ol, last_cursor);
+        }
+        else if (std::string(method) == "circle")
+        {
+            cairo_draw_circle(ol, last_cursor);
+	}
+    }
+
+    void pointer_moved(wlr_event_pointer_motion* ev)
+    {
+        auto current_cursor = wf::get_core().get_cursor_position();
+        auto& ol = get_current_overlay();
+
+        if (std::string(method) == "draw")
```

Comment:
surely there's a more optimal way to do this than comparing *strings* on every pointer event :)

Reply:
Yes, I now only compare strings on init and when method option is changed and convert to enum for use with a switch in motion and unclick handler.